### ISSUE TITLE
Stata integration

### ISSUE THREAD

**Issue by wbuchanan** [meyerjp3/jmetrik-cadence#1]
Not sure if you'd be interested, but I started working on doing a proof of concept integration of jMetrik and Stata.  I know there must be a more efficient way to interface the two programs and I'd be more than willing to do the majority of the coding to do so if I had a better understanding of how the different classes and methods need to be connected.  To see the single model/estimator that is implemented right now, you can visit the repository [StatajMetrik](https://github.com/wbuchanan/StatajMetrik).  

The work flow would basically be something along the lines of someone doing all of their data prep/cleaning/munging in Stata and then calling a function/command in Stata that uses the JavaAPI to pull the data into the JVM and use the classes/methods from jMetrik to estimate the model parameters and then return them to the user in Stata as either additional variables (e.g., for person parameters), matrices (e.g., for item parameters), and/or scalars.

**Comment by meyerjp3** [comment]
Yes, that looks very interesting. I’ve never tried to run Java classes from Stata. You code makes it look fairly straightforward. I have tried it with R, but did not get very far with it. The flow of you design seems reasonable. Have you gotten it to work yet? Let me know if I can guide you through the API. Thanks for taking this on.

 

Patrick

**Comment by wbuchanan** [comment]
I was able to get the JMLE estimator for the Rasch model with binary data to work.  I'm not sure how the commandbuilder package works, but it seems like there should be some way to construct an object to access the data, fit the model to the data, and access the results to pass back into Stata.

**Comment by meyerjp3** [comment]
The commandbuilder package is only for the jMetrik GUI. It mainly defines the jMetrik commands and processes text input from the user. You should be able to run the classes in the psychometrics API without that package. The only benefit to reviewing classes that implement a class in commandbuilder is that it tells you the options that are possible with the API. 

 

You have already found the main entry point for JMLE. There is a similar entry point for MaximumLikelihoodEstimation, and IrtScaleLinking. 

 

Unfortunately, some of the functionality like classical item analysis only exist in parts in the psychometrics library. I have not create an omnibus class that gives you an easy entry point like I have with JMLE and MMLE. It’s on my todo list, but I have not yet done it.

 

If you have gotten JMLE to work with binary items, you should have no trouble getting it to work with polytomous or mixed item types. See the test classes for example of how to instantiate item response model objects and use them along with the data to run the estimation, https://github.com/meyerjp3/psychometrics/blob/master/src/test/java/com/itemanalysis/psychometrics/irt/estimation/JointMaximumLikelihoodEstimationTest.java

**Comment by meyerjp3** [comment]
The commandbuilder package is only for the jMetrik GUI. It mainly defines the jMetrik commands and processes text input from the user. You should be able to run the classes in the psychometrics API without that package. The only benefit to reviewing classes that implement a class in commandbuilder is that it tells you the options that are possible with the API. 

 

You have already found the main entry point for JMLE. There is a similar entry point for MaximumLikelihoodEstimation, and IrtScaleLinking. 

 

Unfortunately, some of the functionality like classical item analysis only exist in parts in the psychometrics library. I have not create an omnibus class that gives you an easy entry point like I have with JMLE and MMLE. It’s on my todo list, but I have not yet done it.

 

If you have gotten JMLE to work with binary items, you should have no trouble getting it to work with polytomous or mixed item types. See the test classes for example of how to instantiate item response model objects and use that along with the data to run the estimation, https://github.com/meyerjp3/psychometrics/blob/master/src/test/java/com/itemanalysis/psychometrics/irt/estimation/JointMaximumLikelihoodEstimationTest.java
